devicetypes/stelpro/stelpro-ki-thermostat.src/stelpro-ki-thermostat.groovy

```diff
@@ -0,0 +1,572 @@
+			log.debug( "Outdoor Temperature: ${tempToSend} C" )
+		}
+		else {
+			tempToSend = weather.current_observation.temp_f
+			log.debug( "Outdoor Temperature: ${tempToSend} F" )
+		}
+		sendEvent( name: 'outsideTemp', value: tempToSend )
+		quickSetOutTemp(tempToSend)
+	}
+}
+
+def scheduledUpdateWeather() {
+	def actions = updateWeather()
+
+	if (actions) {
+		sendHubCommand(actions)
+	}
+}
+
+// Command Implementations
+
+/**
+ * PING is used by Device-Watch in attempt to reach the Device
+ **/
+def ping() {
+	zwave.sensorMultilevelV3.sensorMultilevelGet().format()
+}
+
+def poll() {
+	delayBetween([
+			updateWeather(),
+			zwave.thermostatOperatingStateV1.thermostatOperatingStateGet().format(),
+			zwave.thermostatModeV2.thermostatModeGet().format(),
+			zwave.thermostatSetpointV2.thermostatSetpointGet(setpointType: 1).format(),
+			zwave.sensorMultilevelV3.sensorMultilevelGet().format() // current temperature
+		], 100)
+}
+
+// Event Generation
+def zwaveEvent(physicalgraph.zwave.commands.thermostatsetpointv2.ThermostatSetpointReport cmd) {
+	def cmdScale = cmd.scale == 1 ? "F" : "C"
+	def temp;
+	float tempfloat;
+	def map = [:]
+	if (cmd.scaledValue >= 327) {
+		return [:]
+	}
+	else {
+		temp = convertTemperatureIfNeeded(cmd.scaledValue, cmdScale, cmd.precision)
+		tempfloat = (Math.round(temp.toFloat() * 2)) / 2
+		map.value = tempfloat
+	}
+	map.unit = getTemperatureScale()
+	map.displayed = false
+	switch (cmd.setpointType) {
+		case 1:
+			map.name = "heatingSetpoint"
+			break;
+		default:
+			return [:]
+	}
+	// So we can respond with same format
+	state.size = cmd.size
+	state.scale = cmd.scale
+	state.precision = cmd.precision
+
+	map
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.sensormultilevelv3.SensorMultilevelReport cmd) {
+	def temp
+	float tempfloat
+	def format
+	def map = [:]
+	if (cmd.sensorType == 1) {
+		map.value = convertTemperatureIfNeeded(cmd.scaledSensorValue, cmd.scale == 1 ? "F" : "C", cmd.precision)
+		map.unit = getTemperatureScale()
+		map.name = "temperature"
+
+		temp = map.value
+		if (temp == "32765") {			//0x7FFD
+			map.name = "temperatureAlarm"
+			map.value = "freeze"
+			map.unit = ""
+		}
+		else if (temp == "32767") {		//0x7FFF
+			map.name = "temperatureAlarm"
+			map.value = "heat"
+			map.unit = ""
+		}
+		else if (temp == "-32768"){		//0x8000
+			map.name = "temperatureAlarm"
+			map.value = "cleared"
+			map.unit = ""
+		}
+		else {
+			tempfloat = (Math.round(temp.toFloat() * 2)) / 2
+			map.value = tempfloat
+		}
+
+	}
+	else if (cmd.sensorType == 5) {
+		map.value = cmd.scaledSensorValue
+		map.unit = "%"
+		map.name = "humidity"
+	}
+
+	map
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.thermostatoperatingstatev1.ThermostatOperatingStateReport cmd) {
+	def map = [:]
+	switch (cmd.operatingState) {
+		case physicalgraph.zwave.commands.thermostatoperatingstatev1.ThermostatOperatingStateReport.OPERATING_STATE_IDLE:
+		map.value = "idle"
```

Comment:
needs a tab (or a refactoring)

something like a dict of operating state -> string